src/Http/Middleware/SessionCsrfProtectionMiddleware.php

```diff
@@ -0,0 +1,217 @@
+
+        if ($method === 'GET') {
+            return $handler->handle($request);
+        }
+
+        if ($hasData) {
+            $this->validateToken($request, $session);
+            $request = $this->unsetTokenField($request);
+        }
+
+        return $handler->handle($request);
+    }
+
+    /**
+     * Set callback for allowing to skip token check for particular request.
+     *
+     * The callback will receive request instance as argument and must return
+     * `true` if you want to skip token check for the current request.
+     *
+     * @param callable $callback A callable.
+     * @return $this
+     */
+    public function skipCheckCallback(callable $callback)
+    {
+        $this->skipCheckCallback = $callback;
+
+        return $this;
+    }
+
+    /**
+     * Remove CSRF protection token from request data.
+     *
+     * This ensures that the token does not cause failures during
+     * form tampering protection.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request object.
+     * @return \Psr\Http\Message\ServerRequestInterface
+     */
+    protected function unsetTokenField(ServerRequestInterface $request): ServerRequestInterface
+    {
+        $body = $request->getParsedBody();
+        if (is_array($body)) {
+            unset($body[$this->_config['field']]);
+            $request = $request->withParsedBody($body);
+        }
+
+        return $request;
+    }
+
+    /**
+     * Create a new token to be used for CSRF protection
+     *
+     * This token is a simple unique random value as the compare
+     * value is stored in the session where it cannot be tampered with.
+     *
+     * @return string
+     */
+    public function createToken(): string
+    {
+        return Security::randomString(static::TOKEN_VALUE_LENGTH);
+    }
+
+    /**
+     * Validate the request data against the cookie token.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request to validate against.
+     * @param \Cake\Http\Session $session The session instance.
+     * @return void
+     * @throws \Cake\Http\Exception\InvalidCsrfTokenException When the CSRF token is invalid or missing.
+     */
+    protected function validateToken(ServerRequestInterface $request, Session $session): void
+    {
+        $token = $session->read($this->_config['key']);
+        if (!$token || !is_string($token)) {
+            throw new InvalidCsrfTokenException(__d('cake', 'Missing or incorrect CSRF session key'));
+        }
+
+        $body = $request->getParsedBody();
+        if (is_array($body) || $body instanceof ArrayAccess) {
+            $post = (string)Hash::get($body, $this->_config['field']);
+            if (hash_equals($post, $token)) {
+                return;
+            }
+        }
+
+        $header = $request->getHeaderLine('X-CSRF-Token');
+        if (hash_equals($header, $token)) {
+            return;
+        }
+
+        throw new InvalidCsrfTokenException(__d(
```

Comment:
I thought we dont translate internal messages anymore? Or are those visible to the end user on the page?

Reply:
These can show up in error pages. The other CSRF middleware also has these translated.